KAFKA-3443 [Kafka Streams] added support for subscribing to topics via regex

bbejeck opened the pull request.

bbejeck committed:
KAFKA-3443 added support for subscribing to topics via regex in addSource() and stream()

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/KStreamBuilder.java:
```diff
@@ -55,6 +56,21 @@ public KStreamBuilder() {
+     * Create a {@link KStream} instance from the specified Pattern.
+     * The default deserializers specified in the config are used.
+     * <p>
+     * If multiple topics are specified there are nor ordering guaranteed for records from different topics.
```
nit: "If multiple topics are matched by the specified pattern, the created stream will read data from all of them, and there is no ordering guarantee between records from different topics." Ditto below.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/KStreamBuilder.java:
```diff
@@ -75,6 +91,27 @@ public KStreamBuilder() {
+     * <p>
+     * If multiple topics match the Pattern there are nor ordering guaranteed for records from different topics.
+     *
+     * @param keySerde  key serde used to read this source {@link KStream},
```
Mention that this serde is used for all topics, hence all possibly matched topics should have the same message key-value data format.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/TopologyBuilder.java:
```diff
@@ -112,12 +119,14 @@ public ProcessorNode build(String applicationId) {
+        private SourceNodeFactory(String name, String[] topics, Pattern pattern, Deserializer keyDeserializer, Deserializer valDeserializer) {
             super(name);
-            this.topics = topics.clone();
+            this.topics = topics != null ? topics.clone() : null;
```
Validate that only one of `topics` and `pattern` should be specified.

EDIT: I realized that both of them can be not null, so never mind. But I'm wondering where is this variable used?

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/TopologyBuilder.java:
```diff
@@ -227,21 +253,72 @@ public final TopologyBuilder addSource(String name, Deserializer keyDeserializer
             throw new TopologyBuilderException("Processor " + name + " is already added.");
 
+        if (topicPattern != null) {
+            throw new TopologyBuilderException("Subscription to topics, partitions and pattern are mutually exclusive");
```
We do not need to mention "partition" here since it is supposed to be abstracted from users, ditto below.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/TopologyBuilder.java:
```diff
@@ -227,21 +253,72 @@ public final TopologyBuilder addSource(String name, Deserializer keyDeserializer
     }
 
     /**
+     * Add a new source that consumes from topics matching the given pattern
```
We should probably mention that due to consumer design restriction, currently we only allow one stream throughout the topology to be created from regex patterns.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/TopologyBuilder.java:
```diff
@@ -227,21 +253,72 @@ public final TopologyBuilder addSource(String name, Deserializer keyDeserializer
+        }
+
+        if (this.topicPattern != null) {
+            throw new TopologyBuilderException("Pattern source already defined");
```
nit: "Only one source node with regex subscription pattern can be defined in the topology, and there is an existing pattern: {}".

EDIT: actually, could we union multiple patterns with `[pattern1] [pattern2] ..`? https://docs.oracle.com/javase/7/docs/api/java/util/regex/Pattern.html

Note that if we do this, we need to deserialize data differently based on which sub-patterns its topic name matches to.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/TopologyBuilder.java:
```diff
@@ -473,6 +550,21 @@ public final TopologyBuilder addInternalTopic(String topicName) {
+     *
+     * @param topics  the topics subscribed to from matching the Pattern given
+     */
+    public synchronized void updatePatternNodeTopics(String[] topics) {
```
Could we not make it public since `TopologyBuilder` is a public API?

bbejeck replied:
> Could we not make it public since TopologyBuilder is a public API?

@guozhangwang - I wasn't comfortable with the method being public, but the only point at which topics matching the given pattern are visible to the TopologyBuilder is in the StreamPartitionAssignor class (line 264).  The only other option I came up with was to move StreamPartitionAssignor up out of the internals package and make the method package private, but I don't think that is a good idea.

bbejeck replied:
> Could we not make it public since TopologyBuilder is a public API?

@guozhangwang - I wasn't comfortable with the method being public, but the only point at which topics matching the given pattern are visible to the TopologyBuilder is in the StreamPartitionAssignor class (line 264). The only other option I came up with was to move StreamPartitionAssignor up out of the internals package and make the method package private, but I don't think that is a good idea.

EDIT: I'll come up with another approach.

guozhangwang commented:
Thanks for your patch @bbejeck . One general comment:

In `KafkaConsumer` we are restricting users to exclusively use either topics or pattern subscription. But in `KafkaStreams` it makes less sense to require them doing so. What would be better is that, for example:

```
stream1 = builder.stream(..., "topic1");
stream2 = builder.stream(..., pattern2);
stream3 = builder.stream(..., pattern3);
```

Then the streams library should be able to subscribe to `topic1 | [pattern2] | [pattern3]` (it may not be the correct syntax, but just the idea) as a single pattern subscription in the underlying consumer, and also remember a finer mapping from topic | pattern to the de-/serializers used upon fetching data and putting them to the buffer.

bbejeck commented:
> Then the streams library should be able to subscribe to topic1 | [pattern2] | [pattern3]....

@guozhangwang. 
  Agreed, changes are underway.  Thanks for the feedback.

bbejeck committed:
KAFKA-3443 updates per suggested changes, support for multiple regex and named topic sources.

bbejeck commented:
@guozhangwang 

Changes made per comments.

bbejeck commented:
I don't think the changes from this PR are responsible for the build failure.

bbejeck committed:
KAFKA-3443 updated javadoc to specify topics matched by regex should share key-value data format.

bbejeck commented:
Again, I don't think the pushed changes are responsible for the failure.  Tests for the streams project run fine locally, not sure what the error is from.

guozhangwang commented:
Hmm, this is quite interesting..

```
pure virtual method called
terminate called without an active exception
:streams:test FAILED
```

We have seen this failure before, which I believe is due to RocksDB instances not closed in time, and I thought it has been resolved in https://github.com/apache/kafka/pull/1258, since then I have not seen it again.

@bbejeck I have not looked into the patch in details, but could you think of any chance that the changes could affects how we close and dispose any RocksDB objects?

bbejeck commented:
Well that's certainly possible, but the tests run fine locally.  I didn't think the changes would have caused the failures because the net effect is mostly the same as the original PR, I just took a different approach.  I'll update and re-run the tests.  

EDIT: I just updated/rebased and all the streams tests pass locally.  I'm running on Mac OS X   10.10.5.  I'll continue to look and see if I can find something.

guozhangwang commented:
@bbejeck I re-ran the unit test with your patch and it passed locally as well, so I think this may be due to some corner cases still unresolved from #1258 , will investigate further separately by enhancing some logging information.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/TopologyBuilder.java:
```diff
@@ -111,15 +118,34 @@ public ProcessorNode build(String applicationId) {
+            this.pattern = pattern;
             this.keyDeserializer = keyDeserializer;
             this.valDeserializer = valDeserializer;
+            this.subscriptionUpdates = subscriptionUpdates;
```
Seems that we always pass the same `subscriptionUpdates` object to all the `SourceNodeFactory`, could we instead overload `getTopics` with `getTopics(subscriptionUpdates.getUpdates())`, and trigger either one of the functions based on whether it contains the pattern or not?

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/TopologyBuilder.java:
```diff
@@ -231,17 +274,67 @@ public final TopologyBuilder addSource(String name, Deserializer keyDeserializer
+        }
+
+        nodeToSourcePatterns.put(name, topicPattern);
+        nodeToSourceTopics.put(name, new String[]{});
```
Why do we need to put a dummy entry into `nodeToSourceTopics`?

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java:
```diff
@@ -255,6 +259,14 @@ public Subscription subscription(Set<String> topics) {
             state.capacity = state.capacity + 1d;
         }
 
+        // Source nodes are build ahead of time, but it's at this
```
nit: we do not need to capitalize first letter for in function comments. Also how about changing the comment content to "update the topic groups with the returned subscription list for regex pattern subscriptions".

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java:
```diff
@@ -255,6 +259,14 @@ public Subscription subscription(Set<String> topics) {
+        // point we see the topics subscribed to
+        if (streamThread.builder.sourceTopicPattern() != null) {
+            SubscriptionUpdates subscriptionUpdates = streamThread.builder.getSubscriptionUpdates();
+            subscriptionUpdates.addTopicUpdates(regexSubscribedTopics);
```
Should we "replace" the topic list instead of just adding to the list? Also, how about changing the function name to "updateTopics" since we should not "add" but replace here, right?

bbejeck committed:
KAFKA-3443 Changes made per review comments

KAFKA-3735: Dispose all RocksObejcts upon completeness

Author: Guozhang Wang <wangguoz@gmail.com>

Reviewers: Roger Hoover, Eno Thereska, Ismael Juma

Closes #1411 from guozhangwang/K3735-dispose-rocksobject

MINOR: Specify keyalg RSA for SSL key generation

Author: Sriharsha Chintalapani <harsha@hortonworks.com>

Reviewers: Ismael Juma <ismael@juma.me.uk>

Closes #1416 from harshach/ssl-doc-fix

KAFKA-3747; Close `RecordBatch.records` when append to batch fails

With this change, `test_producer_throughput` with message_size=10000, compression_type=snappy and a snappy buffer size of 32k can be executed in a heap of 192m in a local environment (768m is needed without this change).

Author: Ismael Juma <ismael@juma.me.uk>

Reviewers: Guozhang Wang <wangguoz@gmail.com>

Closes #1418 from ijuma/kafka-3747-close-record-batch-when-append-fails

MINOR: Fix documentation table of contents and `BLOCK_ON_BUFFER_FULL_DOC`

Author: Ismael Juma <ismael@juma.me.uk>

Reviewers: Gwen Shapira

Closes #1423 from ijuma/minor-doc-fixes

Minor: Fix ps command example in docs

Process grep command has been updated. Previous "ps  | grep server-1.properties"  command is showing nothing.

Author: Satendra Kumar <satendra@knoldus.com>

Reviewers: Gwen Shapira

Closes #1386 from satendrakumar06/patch-1

KAFKA-3683; Add file descriptor recommendation to ops guide

Adding sizing recommendations for file descriptors to the ops guide.

Author: Dustin Cote <dustin@confluent.io>
Author: Dustin Cote <dustin@dustins-mbp.attlocal.net>

Reviewers: Gwen Shapira

Closes #1353 from cotedm/KAFKA-3683 and squashes the following commits:

8120318 [Dustin Cote] Adding file descriptor sizing recommendations
0908aa9 [Dustin Cote] Merge https://github.com/apache/kafka into trunk
32315e4 [Dustin Cote] Merge branch 'trunk' of https://github.com/cotedm/kafka into trunk
13309ed [Dustin Cote] Update links for new consumer API
4dcffc1 [Dustin Cote] Update links for new consumer API

MINOR: Add virtual env to Kafka system test README.md

Author: Liquan Pei <liquanpei@gmail.com>

Reviewers: Gwen Shapira

Closes #1346 from Ishiihara/add-venv

MINOR: Removed 1/2 of the hardcoded sleeps in Streams

Author: Eno Thereska <eno.thereska@gmail.com>

Reviewers: Guozhang Wang <wangguoz@gmail.com>, Ismael Juma <ismael@juma.me.uk>

Closes #1422 from enothereska/minor-integration-timeout2

KAFKA-3732: Add an auto accept option to kafka-acls.sh

Added a new argument to AclCommand: --yes. When set, automatically answer yes to prompts

Author: Mickael Maison <mickael.maison@gmail.com>

Reviewers: Gwen Shapira

Closes #1406 from mimaison/KAFKA-3732

KAFKA-3718; propagate all KafkaConfig __consumer_offsets configs to OffsetConfig instantiation

Kafka has two configurable compression codecs: the one used by the client (source codec) and the one finally used when storing into the log (target codec). The target codec defaults to KafkaConfig.compressionType and can be dynamically configured through zookeeper.

The GroupCoordinator appends group membership information into the __consumer_offsets topic by:
1. making a message with group membership information
2. making a MessageSet with the single message compressed with the source codec
3. doing a log.append on the MessageSet

Without this patch, KafkaConfig.offsetsTopicCompressionCodec doesn't get propagated to OffsetConfig instantiation, so GroupMetadataManager uses a source codec of NoCompressionCodec when making the MessageSet. Let's say we have enough group information such that the message formed exceeds KafkaConfig.messageMaxBytes before compression but would fall below the threshold after compression using our source codec. Even if we had dynamically configured __consumer_offsets with our favorite compression codec, the log.append will throw RecordTooLargeException during analyzeAndValidateMessageSet since the message was unexpectedly uncompressed instead of having been compressed with the source codec defined by KafkaConfig.offsetsTopicCompressionCodec.

Author: Onur Karaman <okaraman@linkedin.com>

Reviewers: Manikumar Reddy <manikumar.reddy@gmail.com>, Jason Gustafson <jason@confluent.io>, Ismael Juma <ismael@juma.me.uk>

Closes #1394 from onurkaraman/KAFKA-3718

Setting broker state as running after publishing to ZK

junrao

Currently, the broker state is set to running before it registers itself in ZooKeeper.  This is too early in the broker lifecycle.  If clients use the broker state as an indicator that the broker is ready to accept requests, they will get errors.  This change is to delay setting the broker state to running until it's registered in ZK.

Author: Roger Hoover <roger.hoover@gmail.com>

Reviewers: Jun Rao <junrao@gmail.com>

Closes #1426 from theduderog/broker-running-after-zk

MINOR: Use `--force` instead of `--yes` in `AclCommand`

To be consistent with `ConfigCommand` and `TopicCommand`.

No release includes this option yet, so we can simply change it.

Author: Ismael Juma <ismael@juma.me.uk>

Reviewers: Mickael Maison, Grant Henke

Closes #1430 from ijuma/use-force-instead-of-yes-in-acl-command and squashes the following commits:

bdf3a57 [Ismael Juma] Update `AclCommandTest`
78b8467 [Ismael Juma] Change variable name to `forceOpt`
0bb27af [Ismael Juma] Use `--force` instead of `--yes` in `AclCommand`

MINOR: Fix wrong comments

Author: Yukun Guo <gyk.net@gmail.com>

Reviewers: Gwen Shapira

Closes #1198 from gyk/fix-comment

KAFKA-3723: Cannot change size of schema cache for JSON converter

Author: Christian Posta <christian.posta@gmail.com>

Reviewers: Ewen Cheslack-Postava <ewen@confluent.io>

Closes #1401 from christian-posta/ceposta-connect-class-cast-error

KAFKA-3710: MemoryOffsetBackingStore shutdown

ExecutorService needs to be shutdown on close, lest a zombie thread
prevent clean shutdown.

ewencp

Author: Peter Davis <peter.davis@expeditors.com>

Reviewers: Liquan Pei <liquanpei@gmail.com>, Ewen Cheslack-Postava <ewen@confluent.io>

Closes #1383 from davispw/KAFKA-3710

MINOR: Delete unused code in FileStreamSourceTask

Author: leisore <leisore@gmail.com>

Reviewers: Ismael Juma <ismael@juma.me.uk>, Ewen Cheslack-Postava <ewen@confluent.io>

Closes #1433 from leisore/master

KAFKA-3749; fix "BOOSTRAP_SERVERS_DOC" typo

Author: manuzhang <owenzhang1990@gmail.com>

Reviewers: Guozhang Wang <wangguoz@gmail.com>, Ewen Cheslack-Postava <ewen@confluent.io>, Ismael Juma <ismael@juma.me.uk>

Closes #1420 from manuzhang/KAFKA-3749

MINOR: Fix tracing in KafkaApis.handle()

requestObj() returns null for the o.a.k.c.requests objects so use header() for these.

Once all the requests will have been replaced by o.a.k.c.requests objects, we should be able to clean that up, but in the meantime it's useful to trace both.

Author: Mickael Maison <mickael.maison@gmail.com>

Reviewers: Ismael Juma <ismael@juma.me.uk>

Closes #1435 from mimaison/kafkaapis_trace

MINOR: Fix a couple of scaladoc typos

Author: Vahid Hashemian <vahidhashemian@us.ibm.com>

Reviewers: Ismael Juma <ismael@juma.me.uk>

Closes #1440 from vahidhashemian/typo06/fix_typos_in_code_comments

KAFKA-3682; ArrayIndexOutOfBoundsException thrown by

SkimpyOffsetMap.get() when full

Limited number of attempts to number of map slots after the internal
positionOf() goes into linear search mode.
Added unit test

Co-developed with mimaison

Author: edoardo <ecomar@uk.ibm.com>

Reviewers: Jun Rao <junrao@gmail.com>

Closes #1352 from edoardocomar/KAFKA-3682

KAFKA-3678: Removed sleep from streams integration tests

Author: Eno Thereska <eno.thereska@gmail.com>

Reviewers: Guozhang Wang <wangguoz@gmail.com>

Closes #1439 from enothereska/KAFKA-3678-timeouts1

KAFKA-3767; Add missing license to connect-test.properties

This address to https://issues.apache.org/jira/browse/KAFKA-3767.

Author: Sasaki Toru <sasakitoa@nttdata.co.jp>

Reviewers: Ismael Juma <ismael@juma.me.uk>

Closes #1443 from sasakitoa/test_failure_no_license

KAFKA-3158; ConsumerGroupCommand should tell whether group is actually dead

This patch fix differentiates between when a consumer group is rebalancing or dead and reports the appropriate error message.

Author: Ishita Mandhan <imandha@us.ibm.com>

Reviewers: Vahid Hashemian <vahidhashemian@us.ibm.com>, Jason Gustafson <jason@confluent.io>, Ismael Juma <ismael@juma.me.uk>

Closes #1429 from imandhan/KAFKA-3158

KAFKA-3765; Kafka Code style corrections

Removed explicit returns, not needed parentheses, corrected variables, removed unused imports
Using isEmpty/nonEmpty  instead of size check, using head, flatmap instead of map-flatten

Author: Joshi <rekhajoshm@gmail.com>
Author: Rekha Joshi <rekhajoshm@gmail.com>

Reviewers: Ismael Juma <ismael@juma.me.uk>

Closes #1442 from rekhajoshm/KAFKA-3765

MINOR: Remove synchronized as the tasks are executed sequentially

Author: Liquan Pei <liquanpei@gmail.com>

Reviewers: Ewen Cheslack-Postava <ewen@confluent.io>

Closes #1441 from Ishiihara/remove-synchronized

MINOR: Avoid trace logging computation in `checkEnoughReplicasReachOffset`

`numAcks` is only used in the `trace` logging statement so it should be a `def` instead of a `val`. Also took the chance to improve the code and documentation a little.

Author: Ismael Juma <ismael@juma.me.uk>

Reviewers: Guozhang Wang <wangguoz@gmail.com>, Ewen Cheslack-Postava <ewen@confluent.io>

Closes #1449 from ijuma/minor-avoid-trace-logging-computation-in-partition

bbejeck commented:
@guozhangwang
 made changes per review comments.  Looks like my rebase/squash was unsuccessful let me know if you want me to close this PR and issue another one.

guozhangwang commented:
@bbejeck Could you rebase? Seems some other commits are squashed in this PR.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/TopologyBuilder.java:
```diff
@@ -504,9 +593,18 @@ private void connectProcessorAndStateStore(String processorName, String stateSto
     public Map<Integer, TopicsInfo> topicGroups(String applicationId) {
         Map<Integer, TopicsInfo> topicGroups = new HashMap<>();
 
+
```
nit: add some comments explaining about why we need to update the `nodeToSourceTopics` here.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java:
```diff
@@ -255,6 +259,13 @@ public Subscription subscription(Set<String> topics) {
+        if (streamThread.builder.sourceTopicPattern() != null) {
+            SubscriptionUpdates subscriptionUpdates = streamThread.builder.getSubscriptionUpdates();
+            subscriptionUpdates.updateTopics(regexSubscribedTopics);
+            this.topicGroups = streamThread.builder.topicGroups(streamThread.applicationId);
```
Since now we will update the `this.topicGroups` if there is a sourceTopicPattern, I'm thinking if now we could remove `this.topicGroups = streamThread.builder.topicGroups(streamThread.applicationId);` in the `configure` function of `StreamPartitionAssignor` (line 121)?

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/TopologyBuilder.java:
```diff
@@ -207,6 +230,23 @@ public final TopologyBuilder addSource(String name, String... topics) {
+
+    /**
+     * Add a new source that consumes from topics matching the given pattern
+     * and forwards the records to child processor and/or sink nodes.
```
nit: forward.

EDIT: I realized it may be inherited from the overloaded function, we could fix both.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java:
```diff
@@ -228,12 +228,16 @@ public Subscription subscription(Set<String> topics) {
             Subscription subscription = entry.getValue();
 
+            if (streamThread.builder.sourceTopicPattern() != null) {
+                regexSubscribedTopics.addAll(subscription.topics());
```
I think we should replace with the subscription list here as well, instead of incrementally add to it, since for regex subscription some originally matched topics could be deleted from the broker and hence gets removed from the subscription list.

In fact, in this case, do we still need `regexSubscribedTopics`, or could we just do `subscriptionUpdates.updateTopics(subscription.topics())`?

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/TopologyBuilder.java:
```diff
@@ -207,6 +230,23 @@ public final TopologyBuilder addSource(String name, String... topics) {
+     * @param topicPattern regular expression pattern to match Kafka topics that this source is to consume
+     * @return this builder instance so methods can be chained together; never null
+     */
+    public final TopologyBuilder addSource(String name, Pattern topicPattern) {
```
One general question: I think it is possible that a single topic matches more than one pattern, for example (please ignore the syntax flaws):

```
stream1 = builder.stream("a*");
stream2 = builder.stream("ab*");
```

And it is still allowed in the current setting, right?  In this case, in `TopologyBuilder.build` one topic could be mapped to multiple source nodes (so literally when a record for that topic is fetched during the processing, it should logically be passed to multiple source nodes). How that is handled in your patch?

bbejeck replied:
> One general question: I think it is possible that a single topic matches more than one pattern, for example (please ignore the syntax flaws):

@guozhangwang  
 -for this case I added a check that the topic has not already been added as a source topic (`TopologyBuilder` line 141).  In the worse case of a course grained regex matching all topics of a finer grained regex, the source node with the finer grained regex would not be added to the ProcessorTopology, but all topics would still be processed by the source node with the course grained regex.  Is this a valid condition?

EDIT: Turns out this was not a good approach.  What was done instead was to track patterns by topic when constructing the topology and if a topic maps to more than one Pattern a TopologyBuilderException is thrown.

guozhangwang replied:
Sounds good, thanks.

guozhangwang commented:
This new patch looks very promising. Thanks @bbejeck ! I left some more comments.

Could we add some integration tests (please see examples in streams/test/integration) to validate the case that, for example, with three source streams, two with pattern and one with concrete topic list, 1) the consumer subscription reflects all the matched topics, and when 2) some new topics are added matching the pattern, or 3) some existing topics are deleted matching the pattern, they are correctly reflected during the rebalance, i.e. stream partition assignor?

bbejeck commented:
@guozhangwang 
Thanks for the feedback.  I'll get to work on this soon.

bbejeck commented:
> Could you rebase? Seems some other commits are squashed in this PR.

@guozhangwang 
In my fourth commit I neglected to do a push -f after I had rebased the branch. Are you ok removing the 4th commit only?

bbejeck commented:
@guozhangwang 
Made all changes per comments.

bbejeck commented:
> Could we add some integration tests

@guozhangwang 
With respect to the integration tests, I just need to migrate the ones I have locally to the branch.
I have questions regarding adding/deleting topics.   Is the behavior expected to by dynamic? When I add a topic that matches the pattern I need to restart the stream to get the new topic picked up.
And deleting a topic while the stream is running causes the program to crash with NullPointerException @`org.apache.kafka.clients.Metadata.getClusterForCurrentTopics(Metadata.java:257)`. Starting the stream again resets everything and the topic is dropped successfully.

EDIT: I'm looking into fixing these issues now to allow for dynamic behavior

EDIT 2: Fixed the NullPointerException when deleting a topic that was part of the KStreams (the NPE was occurring whether the topic was named or matched by regex).  The partition assignment is removed for the current topics and a new "generation" is created, but the StreamPartitionsAssignor does not kick in to re-assign the partitions  

Log from KafkaServer after topic delete:
`[2016-06-04 21:18:36,556] INFO Deleted log for partition [topic-A,0] in /tmp/kafka-logs/topic-A-0. (kafka.log.LogManager)
[2016-06-04 21:18:36,723] INFO [GroupCoordinator 0]: Preparing to restabilize group regex_id with old generation 4 (kafka.coordinator.GroupCoordinator)
[2016-06-04 21:18:36,723] INFO [GroupCoordinator 0]: Stabilized group regex_id generation 5 (kafka.coordinator.GroupCoordinator)
[2016-06-04 21:18:36,725] INFO [GroupCoordinator 0]: Assignment received from leader for group regex_id for generation 5 (kafka.coordinator.GroupCoordinator)
[2016-06-04 21:23:36,851] INFO [GroupCoordinator 0]: Preparing to restabilize group regex_id with old generation 5 (kafka.coordinator.GroupCoordinator)
[2016-06-04 21:23:36,851] INFO [GroupCoordinator 0]: Stabilized group regex_id generation 6 (kafka.coordinator.GroupCoordinator)
[2016-06-04 21:23:36,853] INFO [GroupCoordinator 0]: Assignment received from leader for group regex_id for generation 6 (kafka.coordinator.GroupCoordinator)
[2016-06-04 21:26:04,998] INFO [Group Metadata Manager on Broker 0]: Removed 0 expired offsets in 0 milliseconds. (kafka.coordinator.GroupMetadataManager)`

Log from KStream after topic delete:
`[2016-06-04 21:18:36,721] DEBUG Updated cluster metadata version 4 to Cluster(nodes = [oddball.home:9092 (id: 0 rack: null)], partitions = [Partition(topic = topic-1, partition = 0, leader = 0, replicas = [0,], isr = [0,], Partition(topic = topic-Z, partition = 0, leader = 0, replicas = [0,], isr = [0,], Partition(topic = topic-2, partition = 0, leader = 0, replicas = [0,], isr = [0,], Partition(topic = topic-B, partition = 0, leader = 0, replicas = [0,], isr = [0,], Partition(topic = topic-C, partition = 0, leader = 0, replicas = [0,], isr = [0,]]) (org.apache.kafka.clients.Metadata:181)
[2016-06-04 21:18:36,721] INFO Revoking previously assigned partitions [topic-Z-0, topic-1-0, topic-2-0, topic-B-0, topic-C-0] for group regex_id (org.apache.kafka.clients.consumer.internals.ConsumerCoordinator:280)
[2016-06-04 21:18:36,721] INFO Removing a task 0_0 (org.apache.kafka.streams.processor.internals.StreamThread:614)`

I'll continue to work this...

bbejeck committed:
KAFKA-3443 check for null on partitions list for topic removal when streams running.

bbejeck committed:
KAFKA-3443 rebasing removed commit 4 had merge commits

bbejeck commented:
@guozhangwang 
Added integration test for multiple patterns and concrete topic list, and handling for "overlapping" patterns.

bbejeck commented:
I'm aware of the conflict and will resolve.

bbejeck committed:
KAFKA-3443 removed creation of topicGroups in configure method, added clarifying comments, set subscriptions vs adding

bbejeck commented:
Closing this pull request and reopening a new one
